src/main/java/com/google/sps/servlets/ChatServlet.java

```diff
@@ -0,0 +1,137 @@
+    private static final String TIMESTAMP_PROPERTY = "timestamp";
+    private static final String MESSAGE_KIND = "Message";
+
+  /**
+   * When called, doGet will query all previously posted messages to the chat,
+   * and render the messages into a HTML template.
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Calls query on all entities of type Message
+    Query messageQuery = new Query(MESSAGE_KIND).addSort(TIMESTAMP_PROPERTY, SortDirection.ASCENDING);
+    PreparedQuery preparedMessageQuery = datastore.prepare(messageQuery);
+
+    ImmutableMap<String, ImmutableList<String>> data = getTemplateData(preparedMessageQuery);
+
+    // File path starts in target/portfolio-1
+    SoyFileSet sfs = SoyFileSet
+        .builder()
+        .add(new File("../../src/main/java/templates/chat.soy"))
+        .build();
+    SoyTofu tofu = sfs.compileToTofu();
+
+    final String out = tofu.newRenderer("templates.chat.chatPage").setData(data).render();
+    response.getWriter().println(out);
+  }
+
+/**
+ * Is called by a button on the chat form. doPost takes the message and its
+ * attributes (such as timestamp and user) and stores them into Datastore.
+ */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    final String messageText = (String) request.getParameter(MESSAGE_TEXT_PROPERTY);
+    final long timestamp = System.currentTimeMillis();
+    
+    // Reads API Key and Referer from file 
+    File apiKeyFile = new File("../../keys.txt");
+    Scanner scanner = new Scanner(apiKeyFile);
+    final String API_KEY = scanner.nextLine();
+    final String REFERER = scanner.nextLine();
+    scanner.close();
+
+    final String perspectiveURL = "https://commentanalyzer.googleapis.com/v1alpha1/comments:analyze?key=" + API_KEY;
+
+    Double commentScore = getCommentScore(perspectiveURL, REFERER, messageText);
+
+    // If the comment is not toxic, then post it to Datastore
+    if (commentScore <= 0.85) {
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Entity messageEntity = new Entity(MESSAGE_KIND);
+        messageEntity.setProperty(MESSAGE_TEXT_PROPERTY, messageText);
+        messageEntity.setProperty(TIMESTAMP_PROPERTY, timestamp);
+        datastore.put(messageEntity);
+    } else {
+        // TODO: display a message for the user
+        // Either render a text message directly to 
+        // the page or a Javascript style alert
+    }
+
+    response.sendRedirect("/chat");
+  }
+
+  /**
+   * Iterates through the message query to put all the message text into a list.
+   * A map is then created the pass the list of messages into the template.
+   */
+  public ImmutableMap<String, ImmutableList<String>> getTemplateData(PreparedQuery preparedMessageQuery) {
```

Comment:
I think if these methods aren't going to be used outside of the file, they can be private